### PR TITLE
Fix include paths errors

### PR DESCRIPTION
**Pull request by ekpyron**
Closes #3016.

### CODE REVIEW

**Comment by chriseth** [comment]
Please rebase (the PR seems to contain a merge commit).

**Comment by ekpyron** [comment]
@chriseth Rebased.

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
Please add a changelog entry.

File: Changelog.md [hunk 1]
```diff
@@ -27,6 +27,7 @@ Bugfixes:
  * Type System: Improve error message when attempting to shift by a fractional amount.
  * Type System: Make external library functions accessible.
  * Type System: Prevent encoding of weird types.
+ * Commandline interface: fix error messages for imported files that don't exist
```
Comment (chriseth) [inline Changelog.md]: Please use full sentences, start with capital letters and avoid contractions :)

Comment (ekpyron) [reply Changelog.md]: Ok :-).

**Comment by chriseth** [comment]
Did you remove the changelog entry?

**Comment by ekpyron** [comment]
@chriseth Oh - apparently the changelog entry somehow got lost during rebasing - I restored it.

**Comment by ekpyron** [comment]
There is one issue with the solution in this PR.

Consider the paths (1) ``/allowed_dir/non_existing/../contract.sol`` and (2) ``/allowed_dir/non_existing/../../contract.sol``, assuming that ``/allowed_dir/contract.sol`` exists.

The pre-boost-1.60 implementation will treat (1) and (2) as non-existing (regardless of whether ``/contract.sol`` exists or not), whereas the boost>=1.60 implementation will resolve (1) as existing and treat (2) as outside the allowed directories.

In either case no information from outside the allowed directories is leaked, but the behaviour is not entirely consistent. Unfortunately changing this would require to implement ``boost::filesystem::path::lexically_normal`` for boost <1.60, which in turn would require platform specific code.

**Comment by ekpyron** [comment]
So there are several options:

* Merge this as is, accepting the slight inconsistency of builds with different boost versions.
* Add platform specific ``#ifdef``'s to libdevcore to re-implement boost>=1.60's ``lexically_normal`` (in particular Windows paths have to be treated specially, since ``C:\..`` should not result in the empty path).
* Only use the implementation for boost<1.60 for now and change to ``boost::weakly_canonical`` simultanously for all builds, once we can bump the minimum boost version.
* ...

In any case we should not take too long to merge this, since the current version of ``solc`` *will* leak information about non-allowed directories: if you import a file outside the allowed directories that does not exist, ``boost::canonical`` will throw, resulting in the ``unknown exception``, whereas if it does exist, the error is ``File outside of allowed directories.`` - thus current ``solc`` allows to detect file existence everywhere.

**Comment by ekpyron** [comment]
I added this to be done in ``0.4.22``, since it is a security issue.

**Comment by erak** [comment]
I'd prefer using the current implementation for all builds and then bump the version number simultanously once we can. We have to make sure that our implementation is correct anyways, so why not be consistent here. The platform-specific solution might result in too much overhead since we need to test and maintain the code just to throw it out again soon.

**Review by erak — COMMENTED** [review]
Please add unit tests in any case.

**Comment by ekpyron** [comment]
@bit-shift Unit tests for what? Expected error messages for calling ``solc``? Do we even have a suitable testing infrastructure for that yet?

**Comment by erak** [comment]
@ekpyron Oh sorry, could have been more specific. I thought of some tests for ``weaklyCanonical``, but maybe that's not even necessary.

File: libdevcore/CommonIO.h [hunk 2]
```diff
@@ -57,4 +58,6 @@ std::string toString(_T const& _t)
 	return o.str();
 }
 
+boost::filesystem::path weaklyCanonical(boost::filesystem::path const& _path);
```
Comment (axic) [inline libdevcore/CommonIO.h]: Should this be more reflective that it is `boost::filesystem` related function?

File: solc/CommandLineInterface.cpp [hunk 3]
```diff
@@ -716,16 +716,16 @@ bool CommandLineInterface::processInput()
 			}
 			if (!isAllowed)
 				return ReadCallback::Result{false, "File outside of allowed directories."};
-			else if (!boost::filesystem::exists(path))
```
Comment (axic) [inline solc/CommandLineInterface.cpp:719]: Is not using the `else if` statements here is just a matter of style?

Comment (ekpyron) [reply solc/CommandLineInterface.cpp:719]: Yes, it is. These changes were part of the original PR and I left them in. I might as well remove them, though...

Comment (ekpyron) [reply solc/CommandLineInterface.cpp:719]: IMHO it does look nicer with the changes, though.

Comment (axic) [reply solc/CommandLineInterface.cpp:719]: I don't mind, was just curious.

File: libdevcore/CommonIO.h (hunk 2 above)
Comment (axic) [reply libdevcore/CommonIO.h]: I mean should the name of this function contain `filsystem`, `boost` or both?

Comment (ekpyron) [reply libdevcore/CommonIO.h]: What about a comment, stating that it's a partial re-implementation of boost::filesystem::weakly_canonical and to be replaced with it, once we have boost>=1.60?

Comment (axic) [reply libdevcore/CommonIO.h]: I am just not fond that it is a function name without context. It doesn't mean anything when looking at it in the code.

Comment (ekpyron) [reply libdevcore/CommonIO.h]: What about ``weaklyCanonicalFilesystemPath``?

**Review by axic — APPROVED** [review]

File: libdevcore/CommonIO.h (hunk 2 above)
Comment (axic) [reply libdevcore/CommonIO.h]: Sounds good!

**Comment by ekpyron** [comment]
As far as I'm concerned this can be merged once the tests go through.

**Pull request merged**